common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -0,0 +1,307 @@
+static void HandleReply(DBusPendingCall *pending, void *data);
+
+error::Error DBusClient::InitializeConnection() {
+	DBusError dbus_error;
+	dbus_error_init(&dbus_error);
+	dbus_conn_.reset(dbus_bus_get(DBUS_BUS_SYSTEM, &dbus_error));
+	if (!dbus_conn_) {
+		return MakeError(
+			ConnectionError,
+			string("Failed to get connection to system bus: ") + dbus_error.message + "["
+				+ dbus_error.name + "]");
+	}
+
+	dbus_connection_set_exit_on_disconnect(dbus_conn_.get(), FALSE);
+	if (!dbus_connection_set_watch_functions(
+			dbus_conn_.get(), AddDBusWatch, RemoveDBusWatch, ToggleDBusWatch, this, NULL)) {
+		dbus_conn_.reset();
+		return MakeError(ConnectionError, "Failed to set watch functions");
+	}
+	if (!dbus_connection_set_timeout_functions(
+			dbus_conn_.get(), AddDBusTimeout, RemoveDBusTimeout, ToggleDBusTimeout, this, NULL)) {
+		dbus_conn_.reset();
+		return MakeError(ConnectionError, "Failed to set timeout functions");
+	}
+	dbus_connection_set_dispatch_status_function(dbus_conn_.get(), HandleDispatch, this, NULL);
+
+	return error::NoError;
+}
+
+void FreeHandlerCopy(void *data) {
+	DBusCallReplyHandler *handler = static_cast<DBusCallReplyHandler *>(data);
+	delete handler;
+}
+
+error::Error DBusClient::CallMethod(
+	const string &destination,
+	const string &path,
+	const string &iface,
+	const string &method,
+	DBusCallReplyHandler handler) {
+	// TODO: with a mutex to make this thread-safe?
+	if (!dbus_conn_) {
+		auto err = InitializeConnection();
+		if (err != error::NoError) {
+			return err;
+		}
+	}
+
+	unique_ptr<DBusMessage, decltype(&dbus_message_unref)> dbus_msg {
+		dbus_message_new_method_call(
+			destination.c_str(), path.c_str(), iface.c_str(), method.c_str()),
+		dbus_message_unref};
+	if (!dbus_msg) {
+		return MakeError(MessageError, "Failed to create new message");
+	}
+
+	DBusPendingCall *pending;
+	if (!dbus_connection_send_with_reply(
+			dbus_conn_.get(), dbus_msg.get(), &pending, DBUS_TIMEOUT_USE_DEFAULT)) {
+		return MakeError(MessageError, "Unable to add message to the queue");
+	}
+
+	// We need to create a copy here because we need to make sure the handler,
+	// which might be a lambda, even with captures, will live long enough for
+	// the finished pending call to use it.
+	DBusCallReplyHandler *handler_copy = new DBusCallReplyHandler(handler);
+	if (!dbus_pending_call_set_notify(pending, HandleReply, handler_copy, FreeHandlerCopy)) {
+		return MakeError(MessageError, "Failed to set reply handler");
+	}
+
+	return error::NoError;
+}
+
+void HandleDispatch(DBusConnection *conn, DBusDispatchStatus status, void *data) {
+	DBusClient *client = static_cast<DBusClient *>(data);
+	if (status == DBUS_DISPATCH_DATA_REMAINS) {
+		// This must give other things in the loop a chance to run because
+		// dbus_connection_dispatch() below can cause this to be called again.
+		client->loop_.Post([conn, status]() {
+			while (dbus_connection_get_dispatch_status(conn) == DBUS_DISPATCH_DATA_REMAINS) {
```

Comment:
Why does the first `DBUS_DISPATCH_DATA_REMAINS` cause a `Post()`, but all subsequent ones are done in a loop? Based on your comment, I would have assumed that they would repeatedly be calling `Post()` if necessary.

Reply:
The docs say:
> If the dispatch status is [DBUS_DISPATCH_DATA_REMAINS](https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#ggabf250a32ff740b2a8c99136e0142d8d2adf8e8a594ab338cb9cf7855b19e37e93), then [dbus_connection_dispatch()](https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#ga66ba7df50d75f4bda6b6e942430b81c7) needs to be called to process incoming messages. However, [dbus_connection_dispatch()](https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#ga66ba7df50d75f4bda6b6e942430b81c7) MUST NOT BE CALLED from inside the DBusDispatchStatusFunction. Indeed, almost any reentrancy in this function is a bad idea. Instead, the DBusDispatchStatusFunction should simply save an indication that messages should be dispatched later, when the main loop is re-entered.

And I found this when looking at an implementation combining *libdbus* with *libevent* so I did it the same way. It seems to work fine. I believe we only need the initial `Post()` to make sure that when the `HandleDispatch()` function is called, execution of the caller continues and finishes before the actual dispatch happens.

Reply:
Right, that makes sense.